src/boot/efi/devicetree.c

```diff
@@ -0,0 +1,185 @@
+}
+
+static VOID *devicetree_ptr(struct devicetree_state *state) {
+        return (VOID *)(uintptr_t)state->addr;
+}
+
+static EFI_STATUS devicetree_fixup(struct devicetree_state *state, UINTN len) {
+        EFI_DT_FIXUP_PROTOCOL *fixup;
+        UINTN size;
+        EFI_STATUS err;
+
+        err = LibLocateProtocol((EFI_GUID *)&dt_fixup_guid, (VOID **)&fixup);
+        if (EFI_ERROR(err)) {
+                Print(L"Could not locate device tree fixup protocol, skipping.\n");
+                return EFI_SUCCESS;
+        }
+
+        size = devicetree_allocated(state);
+        err = uefi_call_wrapper(fixup->Fixup, 4, fixup,
+                        devicetree_ptr(state), &size,
+                        EFI_DT_APPLY_FIXUPS | EFI_DT_RESERVE_MEMORY);
+        if (err == EFI_BUFFER_TOO_SMALL) {
+                EFI_PHYSICAL_ADDRESS oldaddr = state->addr;
+                UINTN oldpages = state->pages;
+
+                err = devicetree_allocate(state, size);
+                if (EFI_ERROR(err))
+                        return err;
+
+                CopyMem(devicetree_ptr(state), (VOID *)(uintptr_t)oldaddr, len);
+                err = uefi_call_wrapper(BS->FreePages, 2, oldaddr, oldpages);
+                if (EFI_ERROR(err))
+                        return err;
+
+                size = devicetree_allocated(state);
+                err = uefi_call_wrapper(fixup->Fixup, 4, fixup,
+                                devicetree_ptr(state), &size,
+                                EFI_DT_APPLY_FIXUPS | EFI_DT_RESERVE_MEMORY);
+        }
+        if (EFI_ERROR(err))
+                Print(L"Error applying device tree fixups: %r\n", err);
+
+        return err;
+}
+
+EFI_STATUS devicetree_install(struct devicetree_state *state,
+                EFI_FILE_HANDLE root_dir, CHAR16 *name) {
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE handle = NULL;
+        _cleanup_freepool_ EFI_FILE_INFO *info = NULL;
+        UINTN len;
+        EFI_STATUS err;
+
+        err = devicetree_get(state);
+        if (EFI_ERROR(err))
+                return err;
+
+        err = uefi_call_wrapper(root_dir->Open, 5, root_dir, &handle, name,
+                        EFI_FILE_MODE_READ, EFI_FILE_READ_ONLY);
+        if (EFI_ERROR(err))
+                return err;
+
+        info = LibFileInfo(handle);
+        if (!info)
+                return EFI_OUT_OF_RESOURCES;
+        if (info->FileSize < FDT_V1_SIZE || info->FileSize > 32 * 1024 * 1024)
+                /* 32MB device tree blob doesn't seem right */
+                return EFI_INVALID_PARAMETER;
+
+        len = info->FileSize;
+
+        err = devicetree_allocate(state, len);
+        if (EFI_ERROR(err))
+                return err;
+
+        err = uefi_call_wrapper(handle->Read, 3, handle, &len,
+                        devicetree_ptr(state));
+        if (EFI_ERROR(err)) {
+                Print(L"Error reading %s: %r\n", name, err);
```

Comment:
Either this function should log errors, or the caller should. But probably not both.